contributors/design-proposals/controller-ref.md

```diff
@@ -1,100 +1,304 @@
-
-We will need a secondary orphaning mechanism in case of unclean controller deletion:
-* GarbageCollector will remove `ControllerRef` from objects that no longer points to existing controllers
-
-Controller will adopt (set `Controller` field in the `OwnerReference` that points to it) an object whose labels match its selector iff:
-* there are no `OwnerReferences` with `Controller` set to true in `OwnerReferences` array
-* `DeletionTimestamp` is not set
-and
-* Controller is the first controller that will manage to adopt the Pod from all Controllers that have matching label selector and don't have `DeletionTimestamp` set.
-
-By design there are possible races during adoption if multiple controllers can own a given object.
-
-To prevent re-adoption of an object during deletion the `DeletionTimestamp` will be set when deletion is starting. When a controller has a non-nil `DeletionTimestamp` it won't take any actions except updating its `Status` (in particular it won't adopt any objects).
-
-# Implementation plan (sketch):
-
-* Add API field for `Controller`,
-* Extend `OwnerReference` adoption procedure to set a `Controller` field in one of the owners,
-* Update all affected controllers to respect `ControllerRef`.
-
-Necessary related work:
-* `OwnerReferences` are correctly added/deleted,
-* GarbageCollector removes dangling references,
-* Controllers don't take any meaningful actions when `DeletionTimestamps` is set.
-
-# Considered alternatives
+A ControllerRef is thus defined as an OwnerReference with `Controller=true`.
+Each object may have at most one ControllerRef in its list of OwnerReferences.
+The validator for OwnerReferences lists will fail any update that would violate
+this invariant.
+
+# Behavior
+
+This section summarizes the intended behavior for existing controllers.
+It can also serve as a guide for respecting ControllerRef when writing new
+controllers.
+
+## The Three Laws of Controllers
+
+All controllers that manage collections of objects should obey the following
+rules.
+
+1. **Take ownership**
+
+   A controller must claim *ownership* of any objects it creates by adding a
+   ControllerRef, and may also claim ownership of an object it didn't create,
+   as long as the object has no existing ControllerRef (i.e. it is an *orphan*).
+
+1. **Don't interfere**
+
+   A controller may not take any action (e.g. edit/scale/delete) on an object it
+   does not own, except to [*adopt*](#adoption) the object if allowed by the
+   First Law.
+
+1. **Don't share**
+
+   A controller should not count an object it does not own toward satisfying its
+   desired state (e.g. a certain number of replicas), although it may include
+   the object in plans to achieve its desired state (e.g. through adoption)
+   as long as such plans do not conflict with the First or Second Laws.
+
+## Adoption
+
+If a controller finds an orphaned object that matches its selector, it should
```

Comment:
if there's no controller working off a given TPR, it wouldn't matter if you set the controller field to true or not... no controller would actually be acting on that field, and it would behave just like a normal ownerReference